[v5] use fetch api for strapi.fetch

innerdvations opened the pull request:
### What does it do?

- Removes dependency for node-fetch throughout strapi
- Replaces strapi.fetch object with a native Fetch wrapper instead of a node-fetch wrapper
- refactor any node-fetch <-> fetch usage
- replaces the https-global-proxy with an undici ProxyAgent

Note: the plan was and is to remove all third-party modules and exclusively use the Node core Fetch API for everything related to requests, but since fetch is currently in a state between experimental and stable (that is, Node has removed the experimental flag but not yet declared it stable) we will rely on Undici to fill any gaps since its API is identical. If we discover the fetch API to be too unstable for the v5 release, we will rely fully on undici until it can be removed at a later date.

### Why is it needed?

To have a standardized way of performing requests on the backend in the monorepo instead of a wide variety of libraries being used

### How to test it?

- strapi.fetch should work pretty much as before (except now it wraps fetch instead of node-fetch)
- telemetry should work (had to update some calls that now use fetch directly instead of node-fetch)

### Related issue(s)/PR(s)

Roadmap in upcoming PRs:

For every backend request:

- where strapi object is available, use strapi.fetch
- where strapi object is not available, use fetch API directly (replace http/https, request, axios, etc)

innerdvations commented on packages/core/strapi/src/utils/fetch.ts:
```diff
@@ -1,28 +1,31 @@
       ...options,
-    });
+    };
+    strapi.log.debug(`Fetch request for ${url} with ${stringify(fetchOptions)}`);
```
After playing with this for a while, it feels like too much logging, but I think that's because yarn develop should not include debug messages. The server making a remote request seems like something worth logging at the debug level.

innerdvations commented on packages/core/strapi/src/utils/fetch.ts:
```diff
@@ -1,28 +1,31 @@
 
   if (proxy) {
-    fetch.agent = new HttpsProxyAgent(proxy);
+    strapiFetch.dispatcher = new HttpsProxyAgent(proxy);
```
It seems like it works, but will need to make sure it's QA'd thoroughly.

derrickmehaffy commented:
Building experimental to test against squid: https://github.com/strapi/strapi/actions/runs/6641391357

derrickmehaffy commented:
One more after yarn.lock update: https://github.com/strapi/strapi/actions/runs/6642533551

derrickmehaffy commented:
It's running into the same issue I was talking to @Marc-Roig and @joshuaellis about, that command is causing the problem

innerdvations commented on packages/core/strapi/src/services/webhook-runner.ts:
```diff
@@ -125,7 +125,7 @@ class WebhookRunner {
         'Content-Type': 'application/json',
       },
-      timeout: 10000,
+      signal: AbortSignal.timeout(1000),
```
```suggestion
      signal: AbortSignal.timeout(10000),
```

christiancp100 reviewed (approved):
LGTM

derrickmehaffy commented:
> @derrickmehaffy Experimental builds now! https://github.com/strapi/strapi/actions/runs/6773054724
> 
> Could you test out the global proxies and let me know here? Thanks!

Woooooo yeah I'll try and test this week

derrickmehaffy commented:
@innerdvations Proxy is indeed broken here, I also noticed the API `strapi.fetch()` doesn't work the same as in v4. Typically if I call `strapi.fetch` (not a function) I can get the debug information about the fetch client. Here is an example of it in v4:

![image](https://github.com/strapi/strapi/assets/8593673/7c958143-843b-4d69-932b-c4899bbad8a3)

You can see here I can debug this and double check the proxy is set properly. In the v5 experimental I cannot and it's just a function:

![image](https://github.com/strapi/strapi/assets/8593673/2a781169-4516-4f03-b6cc-99f630e68f6c)

The requests you see in this log aren't from the v5 app, it's left over from my testing of the v4 to confirm it works there.

![image](https://github.com/strapi/strapi/assets/8593673/157e94dc-e713-4b8d-80b7-6d5c05955845)

innerdvations commented:
Ok, thanks a lot! If people were using that debug info, I see now why it was attached to fetch, so I can revert that part so it's still available.

Back to the drawing board re: the global proxy itself...

innerdvations commented:
@derrickmehaffy 

I've replaced our `https-proxy-agent` with undici's `ProxyAgent` which should be compatible with the Fetch API considering it's just a direct import of undici, and I've added your "agent" back in with the new name "dispatcher" so it's available for debugging.

Incidentally, we'll now allow a full [ProxyAgent Options config object](https://undici.nodejs.org/#/docs/api/ProxyAgent) to be passed in if desired in addition to just a url string, so now all those nice extras like tls options should work too.

derrickmehaffy commented:
Once it's ready @innerdvations can you do a new experimental and tag me and I'll rerun my tests.

innerdvations commented:
@derrickmehaffy Here you go: 0.0.0-experimental.a30568e92be13a6f2611ccd2a82f0d0dce154156

derrickmehaffy commented:
> @derrickmehaffy Here you go: 0.0.0-experimental.a30568e92be13a6f2611ccd2a82f0d0dce154156

Thank you, I'll try to test it this afternoon or on Monday

derrickmehaffy commented:
So good news / bad news:

- Good news is this works pretty much identical to (ish) to v4

- Bad news is not entirely working with SSO (SSO traffic isn't going over the proxy) but that's honestly similar to v4 although some of it still did not all passport packages do it or do it properly. I tested with Discord using admin SSO and it's traffic did not. I still think we can continue as is and address the SSO stuff later.

Example access log from Squid: 
![image](https://github.com/strapi/strapi/assets/8593673/69218250-4afc-410a-a8bc-40c1dff966c4)

innerdvations commented:
So it doesn't make it worse than v4, it just doesn't improve on it? If that's the case, that's fine, I expect the improvements to come later

Convly reviewed (approved):
Code wise, lgtm

derrickmehaffy commented:
> So it doesn't make it worse than v4, it just doesn't improve on it? If that's the case, that's fine, I expect the improvements to come later

Indeed the bug isn't ours it's deeper in passport. I think this PR is fine as is. If you want to further explore the SSO stuff here is a "temp work around" I have to a customer a while back: https://github.com/strapi-support-demo-apps/strapi-example-v4-sso-proxy/blob/main/patches/passport-oauth2%2B1.7.0.patch

derrickmehaffy commented:
To be honest when I discuss this with Marco about improving SSO in V5 at some point I would recommend that we build our own passport packages and have them set as providers much like we do in the other plugins like email and upload but that's a conversation for another day
